### DISCUSSION TITLE
Enum as field type

### DISCUSSION THREAD

**Discussion by runningjustin** [babyfish-ct/jimmer#665] [Q&A]
I have a field in my Entity defined like so:

`@Default("ACTIVE") val status: UserStatus`

UserStatus is an enum that looks like this:

```
enum class UserStatus {
    CREATED,
    ACTIVE,
    INACTIVE,
    BLOCKED
}
```

When I insert an object with this value set, I have no problems. However, if I want to insert with the default value (ACTIVE) I get the following error from the Postgres driver:

```
Caused by: org.postgresql.util.PSQLException: Can't infer the SQL type to use for an instance of com.package.UserType. Use setObject() with an explicit Types value to specify the type to use.
	at org.postgresql.jdbc.PgPreparedStatement.setObject(PgPreparedStatement.java:1076)
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.setObject(HikariProxyPreparedStatement.java)
	at org.babyfish.jimmer.sql.runtime.DefaultExecutor.setParameters(DefaultExecutor.java:44)
```

I have default enum strategy set to NAME, which I think is default anyway. Do I need to add a scalar provider for all my enums or is there something else I should be doing?

Thanks!

**Comment by babyfish-ct** [comment]
Please define user scalar provider

**Comment by runningjustin** [comment]
Thanks. It seems like this is something Jimmer could do automatically and I didn't see anything in the docs regarding the need to do this with enums. I also had to add a DraftInterceptor to set the default values, as it seems my `@Default` values were not used. Here's the final scalar providers and interceptor:
```
class UserTypeScalarProvider : ScalarProvider<UserType, String> {
    override fun toScalar(sqlValue: String): UserType {
        return UserType.valueOf(sqlValue)
    }

    override fun toSql(scalarValue: UserType): String {
        return scalarValue.name
    }
}

class UserStatusScalarProvider : ScalarProvider<UserStatus, String> {
    override fun toScalar(sqlValue: String): UserStatus {
        return UserStatus.valueOf(sqlValue)
    }

    override fun toSql(scalarValue: UserStatus): String {
        return scalarValue.name
    }
}

class UserDraftInterceptor() : DraftInterceptor<User, UserDraft> {
    override fun beforeSave(draft: UserDraft, original: User?) {
        if (original === null) {
            if (!isLoaded(draft, User::status)) draft.status = UserStatus.CREATED
            if (!isLoaded(draft, User::type)) draft.type = UserType.PERSON
        }
    }
}
```

**Comment by babyfish-ct** [comment]
Oh, enum can be handled by Jimmer automatically.

You ware not saying Postgres specified enum.

**Reply by runningjustin** [reply]
No, it's not an Enum in Postgres (don't want to have to update database when adding new values). It's a String in the database and an Enum in my model.

**Comment by babyfish-ct** [comment]
No problem. please view `testSaveDefaultEnum` of https://github.com/babyfish-ct/jimmer/blob/main/project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/mutation/SaveCommandTest.kt

**Reply by runningjustin** [reply]
That still doesn't seem to work. Could it be a difference between the H2 driver and Postgres or possibly that I'm using Hikari for connection pooling? Here's the error I'm seeing which indicates you should be using the 3-parameter signature of setObject.

```
Caused by: org.postgresql.util.PSQLException: Can't infer the SQL type to use for an instance of com.package.UserType. Use setObject() with an explicit Types value to specify the type to use.
	at org.postgresql.jdbc.PgPreparedStatement.setObject(PgPreparedStatement.java:1076)
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.setObject(HikariProxyPreparedStatement.java)
	at org.babyfish.jimmer.sql.runtime.DefaultExecutor.setParameters(DefaultExecutor.java:44)
	at org.babyfish.jimmer.sql.runtime.DefaultExecutor.access$000(DefaultExecutor.java:16)
	at org.babyfish.jimmer.sql.runtime.DefaultExecutor$BatchContextImpl.add(DefaultExecutor.java:187)
```

**Comment by babyfish-ct** [comment]
Please view `testSaveDefaultEnumByPostgres` of https://github.com/babyfish-ct/jimmer/blob/main/project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/mutation/SaveCommandTest.kt

There is no difference between the H2 driver and Postgres, this test passed.

Please check whether your app changes the default scalar provider of enum. If this problem still exists, please provide minimized reproduce project and sql script.
